File: source/algorithm/NQ084/NQ084_improved.py
def is_isomorphic(s, t):
    if len(s) != len(t):
        return False
    s_to_t = {}
    t_to_s = {}
    for c1, c2 in zip(s, t):
        if c1 in s_to_t:
            if s_to_t[c1] != c2:
                return False  # 检查映射是否一致
        else:
            if c2 in t_to_s:
                return False  # 检查是否有重复映射
            s_to_t[c1] = c2
            t_to_s[c2] = c1
    return True

def main():
    import sys
    n = int(sys.stdin.readline())
    for _ in range(n):
        line = sys.stdin.readline().strip()
        s, t = line.split()
        print("true" if is_isomorphic(s, t) else "false")

if __name__ == "__main__":
    main()

File: 兴趣班入门百练讲义/NQ100/NQ084/NQ084_improved.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;
bool isIsomorphic(const string& s, const string& t) {
    if (s.size() != t.size()) return false;
    unordered_map<char, char> s_to_t, t_to_s;
    for (int i = 0; i < s.size(); i++) {
        char c1 = s[i], c2 = t[i];
        if (s_to_t.count(c1)) {
            if (s_to_t[c1] != c2) return false;  // 检查映射是否一致
        } else {
            if (t_to_s.count(c2)) return false;  // 检查是否有重复映射
            s_to_t[c1] = c2;
            t_to_s[c2] = c1;
        }
    }
    return true;
}
int main() {
    int n;
    cin >> n;
    while (n--) {
        string s, t;
        cin >> s >> t;
        cout << (isIsomorphic(s, t) ? "true" : "false") << endl;
    }
    return 0;
}